app/models/person.rb

```diff
@@ -1,9 +1,159 @@
+  # PersonTimeline. If the end user has requested the deletion, the location
+  # owner is notified.
+
+  def self.destroy_relations(options)
+    return unless options['location_id'].present? && options['person_id'].present?
+
+    Email.where(location_id: options['location_id'], person_id: options['person_id']).destroy_all
+    Social.where(location_id: options['location_id'], person_id: options['person_id']).destroy_all
+    Sms.where(location_id: options['location_id'], person_id: options['person_id']).destroy_all
+    PersonTimeline.where(location_id: options['location_id'], person_id: options['person_id']).destroy_all
+
+    return unless options['portal_request']
+    notify_deletion(options)
+  end
+
+  def self.notify_deletion(options)
+    loc = Location.find_by(id: options['location_id'])
+
+    return unless loc && loc.user_id
+    user = User.find_by(id: loc.user_id)
+    return unless user
+
+    @mailer_opts = {
+      email: user.try(:email),
+      location_name: loc.location_name
+    }
+    DataRequestMailer.with(@mailer_opts).delete_request_email.deliver_now
+  end
+
+  ##
+  # Creates an access code for an end user to access their data. The link to the
+  # page(s) is emailed to them.
+
+  def self.create_portal_links_email(email)
+    mailer_data = create_access_codes(get_people_ids(email))
+    return unless mailer_data
+    @mailer_opts = {
+      email: email,
+      mailer_data: mailer_data,
+      url: "https://#{ENV['MIMO_DASHBOARD_URL']}/#/timeline/"
+    }
+    DataRequestMailer.with(@mailer_opts).access_request_email.deliver_now
+  end
+
+  def self.create_access_codes(ids)
+    mailer_data = []
+    week = 60*60*24*7
+    ids.map {|id| mailer_data << set_timeline_portal_code(week, id)}
+    mailer_data
+  end
+
+  def self.set_timeline_portal_code(week, id)
+    # write a new test for this
+    code = SecureRandom.hex
+    REDIS.setex("timelinePortalCode:#{id}", week, code)
+    {id: id, code: code}
+  end
+
+  def self.get_people_ids(email)
+    ids = []
+    Person.where(email: email).map {|person| ids << person.id }
+    Email.where(email: email).map {|email| ids << email.person_id if email.person_id.present? }
+    Social.where(email: email).map {|social| ids << social.person_id if social.person_id.present? }
+    ids.uniq
+  end
+
+  ##
+  # Creates a zip for with a user's Person data, and all of their associated
+  # Email, Social and Sms data, and emails the zip as an attached report.
+
+  def self.download_person_data(opts)
+    person = Person.find_by(id: opts['person_id'])
+    return unless person
+    csvs = [person_csv(person), emails_csv(opts), social_csv(opts), sms_csv(opts)].compact
+    @mailer_opts = {
+      zip: create_zip(csvs, opts),
+      email: opts['email']
+    }
+    DataRequestMailer.with(@mailer_opts).data_download.deliver_now
+    remove_files(csvs, opts)
+  end
+
+  def self.person_csv(person)
+    file_name = "person_#{person.id}_#{SecureRandom.hex(5)}.csv"
```

Comment:
turn into a function also:

def self.person_csv_filename or something and don't set as a var?